readme


localSetupROOT --rootVersion=5.34.14-x86_64-slc5-gcc4.3
python rajiv.py -n 200 -i /hdfs/projects/atlas/2012_d3pd/data12_8TeV.00212619.physics_JetTauEtmiss.merge.NTUP_SMQCD.r4065_p1278_p1344_p1345_tid01136320_00/NTUP_SMQCD.01136320._000001.root.1


to run over grid:

prun --exec 'python rajiv.py -o debo.root -i %IN'  --inDS data12_8TeV.00212619.physics_JetTauEtmiss.merge.NTUP_SMQCD.r4065_p1278_p1344_p1345/ 
--outDS user.subraman.data12_8TeV.00212619.physics_JetTauEtmiss.merge.05_29_2014 
--outputs debo.root --extFile data12_8TeV.periodAllYear_DetStatus-v58-pro14-01_DQDefects-00-00-33_PHYS_StandardGRL_All_Good.xml,libTrigRootAnalysis.so,libGoodRunsLists.so,ApplyJetCalibration-00-03-03/StandAlone/libApplyJetCalibration.so,ApplyJetCalibration-00-03-03/data/CalibrationConfigs/JES_Full2012dataset_Preliminary_Jan13.config,ApplyJetCalibration-00-03-03/data/InsituCalibration/InsituCalibration_Dec21_2012.root 
--nFilesPerJob=1 --rootVer==5.34.14-x86_64-slc5-gcc4.3


ApplyJetCalibration-00-03-02/Changelog
fApplyJetCalibration-00-00-01
	First version. Contains default Rel 16.6 and Rel 17 jet calibrations,
	and the first set of pile-up corrections

ApplyJetCalibration-00-00-02
	Improved documentation, and options to switch betwen MeV, GeV.
	MeV still the default (ATLAS default - sigh...). UseGeV() expects GeV
	as input and returns all corrections in GeV.

ApplyJetCalibration-00-00-03
	Added Makefile.Standalone

ApplyJetCalibration-00-00-04
	Now compatible with root 5.32. Added details on how each calibration
	was derived in doc folder.

ApplyJetCalibration-00-00-05
	Do not use (missing a file)

ApplyJetCalibration-00-00-06
	Added fist version of in-situ eta intercalibration. Not applied by default.
	The eta-intercalibration is to be used by the other in-situ analyses to
	derive the absolute scale for data.

ApplyJetCalibration-00-00-07
	Added updated version of in-situ eta intercalibration. Not applied by default.
	The eta-intercalibration is to be used by the other in-situ analyses to
	derive the absolute scale for data.

ApplyJetCalibration-00-00-08
	Changed structure of calibration settings files such that they should be
	automatically found using RootCore

ApplyJetCalibration-00-00-09, Feb 9, 2012
	Third "final" iteration of eta intercalibration correction.
	Default version used by other in-situ analyses

ApplyJetCalibration-00-00-10
	Added AFII calibration

ApplyJetCalibration-00-00-11
	New option to apply bad channel correction

ApplyJetCalibration-00-00-12
	Dusted off the changelog
	Added first MC12 calibrations, offsets from mswiatlo, csandova
	Several updates to in-situ corrections to data in the meantime

ApplyJetCalibration-00-00-13
	Preliminary 2011 in-situ calbiration
	First version with both absolute (Z+jet, MPF andmultijet) and relative (eta) in-situ calibration

ApplyJetCalibration-00-01-00, March 29, 2012
	Same as 00-00-13, only small update in the plotting macro.

ApplyJetCalibration-00-01-01, April 12, 2012
	Fixed bug in in-situ applying inverse of in-situ calibration 

ApplyJetCalibration-00-01-02, April 20, 2012
	Smoothing applied to calibration in tag -01-01 

ApplyJetCalibration-00-01-03, April 26, 2012
	New iteration of in-situ calibration, only up to 600 GeV (no MJB, frozen beyond 600 GeV)

ApplyJetCalibration-00-01-03-01, Mar 3, 2012
        Fixed missing root file

ApplyJetCalibration-00-01-04, May 10, 2012
        New iteration of in-situ calibration, full pT range (including MJB)
        Fixed memory leaks for settings and axes (thanks K-J. Grahn)

ApplyJetCalibration-00-01-05, May 10, 2012
        Fixed missing file for new iteration of in-situ calibration, full pT range (including MJB)
        Using smoothing Smooth3 (less hard)

ApplyJetCalibration-00-01-06, May 15, 2012
        Preliminary calibration with new iteration of in-situ calibration, full pT range (including MJB)

ApplyJetCalibration-00-01-07, May 15, 2012
        Preliminary calibration for muScan2 samples, all jet collections, all conditions except thresh = 60
        Using smoothing Smooth1 (harder)

ApplyJetCalibration-00-02-05, August 17, 2012
        Introducing R=0.5, 0.7 calibrations in JES_June2012.config. This file also contains updated
        R=0.4,0.6 calibrations as of 00-02-04. Calibrations from mswiatlo, offsets from csandova
        Added ApplyOffset function

ApplyJetCalibration-00-02-06
         Added August2012 updated calibration for 2012 data/MC
         Added AFII calibrations for 2012 data/MC

ApplyJetCalibration-00-02-07, September 14th, 2012
         implemented support for jet mass correction
         added JES, eta and mass corrections for 4 fat jet collections - for 2011 data/MC

ApplyJetCalibration-00-01-08
         added JES, eta and mass corrections for 2 ungroomed fat jet collections - for 2011 data/MC

ApplyJetCalibration-00-02-08, Oct 15 2012
	First implementation of jet area offset correction with residual correction
	TEST VERSION

ApplyJetCalibration-00-02-09, Oct 16 2012
	Second implementation of jet Area offset correction - now completly decoupled
	from previous pile-up correction

ApplyJetCalibration-00-02-10, Oct 18 2012
	jtaenzer - Minor bug fix in jet Area offset  correction - it now passes the jet pT correctly
	jtaenzer - Fixed some warnings coming from the JetCalibrationTool constructor
	jtaenzer - Added athena version.cmt

ApplyJetCalibration-00-02-11, Oct 18 2012
	mswiatlo- corrected August2102 calibration function for one bin in AntiKt6LCTopo jets (eta 3.2-3.3)
	dgillber - Changed jet area correction to use transverse component of area 4-vector instead of full
	  4-vector correction
	dgillber - Updated residual jet area offset correction factors for LC jets, and added new ones for EM jets (from John)

ApplyJetCalibration-00-02-12, Oct 19 2012
        jtaenzer - Added config flag (ApplyFullJetArea4MomentumCorrection) that allows users to apply full 4-vector area correction. It is off by default.

ApplyJetCalibration-00-02-13, November 02 2012
	jtaenzer - Added November2012 config file, new offset and residual offset correction factors from John.
	jtaenzer - Changes in the offset and residual offset correction factors (copy pasted from John):
	           --more inclusive event selection (no longer requiring a certain number of truth jets within the tracking acceptance)
	           --truth isolation cut (all jets are required to be truth matched within the full radius of the jet, but no other truth jets should exist within 2.5 times the radius)
		   --mu has been rescaled (multiplied by 1.11)
	           --NPV has been adjusted to account for the fact that the beamspot in MC is too wide in z
	jtaenzer - Updated the code so it supports the new format of the offset correction factors (it's still backwards compatible).
	jtaenzer - Fixed minor memory leak in JetCalibrationTool::init()


ApplyJetCalibration-00-02-14, November 13 2012

valid.py
print "\n:: Decode cut flow of an L1.5 Forward Study \n"
#from ROOT import *
print ":: Importing ROOT"
from ROOT import TH1D,TFile,TChain
print ":: done"
from EosTools import make_eos_file_list
from Tools import make_file_list

############## setup option parser
from optparse import OptionParser

parser = OptionParser()
parser.add_option("-i", "--input",          help="file to study",                           action="store", type="string",  dest="input",   default=False)
parser.add_option("-f", "--files",          help="files to study",                          action="store", type="string",  dest="files",   default=False)
parser.add_option("-e", "--eos_folder",     help="eos folder to study",                     action="store", type="string",  dest="eos_dir", default=False)
parser.add_option("-t", "--token",          help="grep token",                              action="store", type="string",  dest="token",   default="")
parser.add_option("-n", "--number_of_runs", help="only run over a certain number of runs",  action="store", type="int",     dest="n_runs",  default=False)
parser.add_option("-v", "--verbose_mode",   help="turn on verbose mode",                    action="store_true",            dest="verbose", default=False)
parser.add_option("-g", "--glados",         help="run on glados",                           action="store_true",            dest="glados",  default=False)
parser.add_option("-r", "--runs",           help="only study these runs",                   action="store", type="string",  dest="runs",    default=False)
parser.add_option("-s", "--strip",          help="strip leading zeros from run names",      action="store_true",            dest="strip",   default=False)


(options, args) = parser.parse_args()

print ":: Options"
print "   input:        ",options.input
print "   files:        ",options.files
print "   eos folder:   ",options.eos_dir
print "   token:        ",options.token
print "   n runs:       ",options.n_runs
print "   runs:         ",options.runs
print "   glados:       ",options.glados
print "   strip:        ",options.strip
print "   verbose:      ",options.verbose

assert options.eos_dir or options.input, "no eos folder specified (use -e) and no input file (use -i)"

if options.files:
    file_list = open(options.files,"r").readlines()
    print "Got %i runs from file_list"%(len(file_list))
    runs = []
    for file in file_list:
        #print file
        try:
            run = file.split(".")[1].strip("\n")
        except IndexError:
            run = file.strip("\n")
        if options.strip: run = run[2:]
        runs.append(run)
elif options.runs:
    runs = options.runs.split(",")
    print "Got %i runs from augments"%len(runs)

elif options.input:
    runs = [options.input.split("/").pop()]
runs.sort()
print "Got %i runs"%len(runs)

nevents = len(runs)
if options.n_runs: 
    nevents = int(options.n_runs)
    print "   running on %i of these"%nevents
    
print "Making files list"
if options.input:       file_list = [options.input]
elif options.glados:    file_list = make_file_list(options.eos_dir,grep=options.token,verbose=options.verbose)
else:                   file_list = make_eos_file_list(options.eos_dir,grep=options.token,verbose=options.verbose)
print "Got %i files"%len(file_list)
#print file_list

if options.glados:  prefix = ""
else:               prefix = "root://eosatlas/"

run_information = {}
bad_runs = []
zero_entry_runs = []
used_files = []
for i,run in enumerate(runs):
    if options.runs and (run not in options.runs): continue
    if i == nevents: 
        print "Reached run limit"
        break
    this_file_list = []
    for file in file_list:
        if run in file: 
            this_file_list.append(file)
            used_files.append(file)
        
    #file_list = make_eos_file_list(options.eos_dir,grep=run,verbose=options.verbose)
    print "[%3i/%3i] - got %i files from eos for run: %s"%(i+1,len(runs),len(this_file_list),run)
    if not len(this_file_list): 
        run_information[run] = [0 for i in range(35)]
        continue
    cut_flow = False
    for file in this_file_list: 
        try:            
            if not cut_flow:
                first_root_file = TFile.Open( prefix+file )    
                assert first_root_file.IsOpen()
                if options.verbose: print "   - root file "+file+" is open"
                cut_flow = first_root_file.Get("cut_flow").Clone("compressed_cut_flow")
                if options.verbose: print "    - cut flow[1]: %i"%cut_flow.GetBinContent(1)
                first_chain = TChain("qcd")
                first_chain.Add(prefix+file)
                n_entries =first_chain.GetEntries()
                
                if cut_flow.GetBinContent(1)==0: zero_entry_runs.append(file)
            else:
                root_file = TFile.Open( prefix+file )    
                assert root_file.IsOpen()
                if options.verbose: print "   - root file "+file+" is open"
                this_cut_flow = root_file.Get("cut_flow")
                if options.verbose: print "    - cut flow[1]: %i"%this_cut_flow.GetBinContent(1)
                cut_flow.Add(this_cut_flow)
                chain = TChain("qcd")
                chain.Add(prefix+file)
                n_entries+=chain.GetEntries()
                if this_cut_flow.GetBinContent(1)==0: zero_entry_runs.append(file)
                root_file.Close()    
        except:
            print "Failure on run: %s, file: %s"%(run,file)
            bad_runs.append(file)
            print "   - %i bad runs so far"%len(bad_runs)
            #cut_flow = False
            #assert False
        
    #if not cut_flow: continue
    offset = 1
    if options.verbose: 
        print "\n=== Summary"
        print "    Entries:     %i"%n_entries
        print "    Events:            %i"%cut_flow.GetBinContent(1)
        print "     - corrupt qcd     %i"%cut_flow.GetBinContent(13)
        print "     - corrupt qcdMeta %i"%cut_flow.GetBinContent(14)
        print "    GRL:               %i"%cut_flow.GetBinContent(2)
        print "    Trigger:           %i"%cut_flow.GetBinContent(3)
        print "    LAr:               %i"%cut_flow.GetBinContent(4)
        print "    tile:              %i"%cut_flow.GetBinContent(5)
        print "    ttc:               %i"%cut_flow.GetBinContent(6)
        print "    Vertex:            %i"%cut_flow.GetBinContent(7)
        print "    Jets:              %i"%cut_flow.GetBinContent(8)
        print "       -looser:        %i"%cut_flow.GetBinContent(9)
        print "       -loose:         %i"%cut_flow.GetBinContent(10)
        print "       -medium:        %i"%cut_flow.GetBinContent(11)
        print "       -tight:         %i"%cut_flow.GetBinContent(12)
        print "       -ugly:          %i"%cut_flow.GetBinContent(13)
        print "    Skim:              %i"%cut_flow.GetBinContent(14)
        print "    Truth jets:        %i"%cut_flow.GetBinContent(17)
        #print "    Vertex:            %i"%cut_flow.GetBinContent(5)
        #print "    Jets:              %i"%cut_flow.GetBinContent(6)
        #print "       -looser:        %i"%cut_flow.GetBinContent(7)
        #print "       -loose:         %i"%cut_flow.GetBinContent(8)
        #print "       -medium:        %i"%cut_flow.GetBinContent(9)
        #print "       -tight:         %i"%cut_flow.GetBinContent(10)
        #print "       -ugly:          %i"%cut_flow.GetBinContent(11)
        #print "    Skim:              %i"%cut_flow.GetBinContent(12)
        #print "    Truth jets:        %i"%cut_flow.GetBinContent(15)
    triggers = [
                ### new naming convention triggers
                "EF_j10_a4tc_EFFS",
                "EF_j15_a4tc_EFFS",
                "EF_j20_a4tc_EFFS",
                "EF_j30_a4tc_EFFS",
                "EF_j40_a4tc_EFFS",
                "EF_j55_a4tc_EFFS",
                "EF_j75_a4tc_EFFS",
                "EF_j100_a4tc_EFFS",
                "EF_j135_a4tc_EFFS",
                "EF_j180_a4tc_EFFS",
                "EF_j240_a4tc_EFFS",
                ### old naming convention triggers were used pre-period D, I don't think we'll need these
                ]      
    
    cut_flow_offset = 23 # pre truth offset was 15, was 16 13/11/12
    if options.verbose: print " By physics trigger"
    for t in triggers:
        if options.verbose: print "       - %20s %i"%(t+":",cut_flow.GetBinContent(cut_flow_offset))    
        cut_flow_offset+=1
        
    if options.verbose: print " RAW trigger"
    for t in triggers:
        if options.verbose: print "       - %20s %i"%(t+":",cut_flow.GetBinContent(cut_flow_offset))    
        cut_flow_offset+=1
        
    run_information[run] = []
    for i in range(34):
        run_information[run].append(cut_flow.GetBinContent(i))
        
    run_information[run].append(len(this_file_list))
    
    first_root_file.Close()
    
if len(bad_runs):
    print "\n=== Bad runs"
    for run in bad_runs:
        command = "xrd eosatlas rm %s"%run
        print command
if len(zero_entry_runs):
    print "\n=== Zero entry runs"
    for run in zero_entry_runs:
        command = "xrd eosatlas rm %s"%run
        print command


print "\n=== Full overall summary"
print "%-10s %15s %15s %15s %15s %15s %15s %15s %15s %15s"%("run","Events","GRL", "Trigger", "LAr", "tile", "ttc", "vertex", "Skim", "files")

for i,run in enumerate(runs):
    if i == nevents: break
    print "%-10s %15i %15i %15i %15i %15i %15i %15i %15i %15i"%\
     (run,run_information[run][1],run_information[run][2],run_information[run][3],run_information[run][4],run_information[run][5],run_information[run][6],run_information[run][7],run_information[run][14],run_information[run][34])
    
    
        
print "\n=== Overall summary"
print "%-10s %15s %15s %10s"%("run","seen by skim","skimmed","files")
all = [0,0,0]
for i,run in enumerate(runs):
    if i == nevents: break
    print "%-10s %15i %15i %10i"%(run,run_information[run][1],run_information[run][14],run_information[run][34])
    all[0]+=run_information[run][0]
    all[1]+=run_information[run][1]
    all[2]+=run_information[run][2]


print "\n=== %i of %i files used"%(len(used_files),len(file_list))
print "unused files:"
for file in file_list:
    if file not in used_files: 
        command = "xrd eosatlas rm %s"%file
        print command
        #print "mv %s ../unused/ "%file.split("/").pop()
        #print file.split("/").pop()

